parquet-variant/src/variant.rs

```diff
@@ -0,0 +1,714 @@
+
+        // Read the value bytes from the offsets
+        let variant_value_bytes = slice_from_slice(
+            self.value,
+            first_value_byte + start_field_offset_from_first_value_byte
+                ..first_value_byte + end_field_offset_from_first_value_byte,
+        )?;
+        let variant = Variant::try_new(self.metadata, variant_value_bytes)?;
+        Ok(variant)
+    }
+}
+
+// impl<'m, 'v> Index<usize> for VariantArray<'m, 'v> {
+//     type Output = Variant<'m, 'v>;
+//
+// }
+
+/// Variant value. May contain references to metadata and value
+#[derive(Clone, Debug, Copy, PartialEq)]
+pub enum Variant<'m, 'v> {
+    // TODO: Add types for the rest of the primitive types, once API is agreed upon
+    Null,
+    Int8(i8),
+
+    BooleanTrue,
+    BooleanFalse,
+
+    // Note: only need the *value* buffer
+    String(&'v str),
+    ShortString(&'v str),
+
+    // need both metadata & value
+    Object(VariantObject<'m, 'v>),
+    Array(VariantArray<'m, 'v>),
+}
+
+impl<'m, 'v> Variant<'m, 'v> {
+    /// Parse the buffers and return the appropriate variant.
+    pub fn try_new(metadata: &'m VariantMetadata, value: &'v [u8]) -> Result<Self, ArrowError> {
+        let header = *first_byte_from_slice(value)?;
+        let new_self = match get_basic_type(header)? {
+            VariantBasicType::Primitive => match get_primitive_type(header)? {
+                VariantPrimitiveType::Null => Variant::Null,
+                VariantPrimitiveType::Int8 => Variant::Int8(decoder::decode_int8(value)?),
+                VariantPrimitiveType::BooleanTrue => Variant::BooleanTrue,
+                VariantPrimitiveType::BooleanFalse => Variant::BooleanFalse,
+                // TODO: Add types for the rest, once API is agreed upon
+                VariantPrimitiveType::String => {
+                    Variant::String(decoder::decode_long_string(value)?)
+                }
+            },
+            VariantBasicType::ShortString => {
+                Variant::ShortString(decoder::decode_short_string(value)?)
+            }
+            VariantBasicType::Object => Variant::Object(VariantObject { metadata, value }),
+            VariantBasicType::Array => Variant::Array(VariantArray { metadata, value }),
+        };
+        Ok(new_self)
+    }
+
+    pub fn as_null(&self) -> Option<()> {
+        match self {
+            Variant::Null => Some(()),
+            _ => None,
+        }
+    }
+
+    pub fn as_boolean(&self) -> Option<bool> {
+        match self {
+            Variant::BooleanTrue => Some(true),
+            Variant::BooleanFalse => Some(false),
+            _ => None,
+        }
+    }
+
+    pub fn as_string(&'v self) -> Option<&'v str> {
+        match self {
+            Variant::String(s) | Variant::ShortString(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    pub fn as_int8(&self) -> Option<i8> {
+        match self {
+            Variant::Int8(i) => Some(*i),
```

Comment:
If we add more conversions, it may prove useful to match on dereferenced `self`:
```suggestion
        match *self {
            Variant::Int8(i) => Some(i),
```
(avoids having to dereference in every match arm)